File: src/limitedwip/watchdog/ui/WatchdogStatusBarWidget.java
package limitedwip.watchdog.ui;

import com.intellij.ide.DataManager;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.wm.StatusBar;
import com.intellij.openapi.wm.StatusBarWidget;
import com.intellij.util.Consumer;
import limitedwip.watchdog.components.WatchdogComponent;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.awt.event.MouseEvent;

public class WatchdogStatusBarWidget implements StatusBarWidget {
    private static final String textPrefix = "Change size: ";

    private String text = "";

    @Override public void install(@NotNull StatusBar statusBar) {
    }

    @Override public void dispose() {
    }

    public void showChangeSize(String linesInChange, int maxLinesInChange) {
        text = textPrefix + linesInChange + "/" + maxLinesInChange;
    }

    public void showInitialText(int maxLinesInChange) {
        text = textPrefix + "-/" + maxLinesInChange;
    }

    @Override public WidgetPresentation getPresentation(@NotNull PlatformType type) {
        return new TextPresentation() {
            @NotNull @Override public String getText() {
                return text;
            }

            @NotNull @Deprecated public String getMaxPossibleText() {
                return "";
            }

            @Override public String getTooltipText() {
                return "Shows amount of changed lines in current change list vs change size limit.";
            }

            @Override public Consumer<MouseEvent> getClickConsumer() {
                return new Consumer<MouseEvent>() {
                    @Override public void consume(MouseEvent mouseEvent) {
                        DataContext dataContext = DataManager.getInstance().getDataContext(mouseEvent.getComponent());
                        Project project = PlatformDataKeys.PROJECT.getData(dataContext);
                        if (project == null) return;
                        WatchdogComponent watchdogComponent = project.getComponent(WatchdogComponent.class);
	                    if (watchdogComponent == null) return;

                        watchdogComponent.toggleSkipNotificationsUntilCommit();
                    }
                };
            }

            @Override public float getAlignment() {
                return Component.CENTER_ALIGNMENT;
            }
        };
    }

    @NotNull @Override public String ID() {
        return "LimitedWIP_" + this.getClass().getSimpleName();
    }
}

File: src/limitedwip/watchdog/ui/WatchdogStatusBarWidget.kt
package limitedwip.watchdog.ui

import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.util.registry.Registry
import com.intellij.openapi.wm.StatusBar
import com.intellij.openapi.wm.StatusBarWidget
import com.intellij.openapi.wm.WindowManager
import com.intellij.util.Consumer
import limitedwip.common.FloatingWidget
import limitedwip.common.pluginId
import java.awt.Component
import java.awt.Point
import java.awt.event.MouseEvent

class WatchdogStatusBarWidget(project: Project): StatusBarWidget {
    private val floatingWidget = FloatingWidget(
        projectComponent = WindowManager.getInstance().getIdeFrame(project)?.component,
        point = Point(
            Registry.intValue("limited-wip.watchdog.widget.x"),
            Registry.intValue("limited-wip.watchdog.widget.y")
        )
    )

    private val textPrefix = "Change size: "
    private var text = ""
    private var linesInChange = ""
    private var maxLinesInChange = ""
    var listener: Listener? = null

    fun showChangeSize(linesInChange: String, maxLinesInChange: Int) {
        this.linesInChange = linesInChange
        this.maxLinesInChange = maxLinesInChange.toString()
        text = "$textPrefix$linesInChange/$maxLinesInChange"
        floatingWidget.text = text
    }

    override fun install(statusBar: StatusBar) {}

    override fun dispose() {
        Disposer.dispose(floatingWidget)
    }

    override fun getPresentation() =
        object: StatusBarWidget.TextPresentation {
            override fun getText() = this@WatchdogStatusBarWidget.text
            override fun getTooltipText() = "Change size in lines: $linesInChange; threshold: $maxLinesInChange"
            override fun getClickConsumer() = Consumer<MouseEvent> { listener?.onClick() }
            override fun getAlignment() = Component.CENTER_ALIGNMENT
        }

    override fun ID() = pluginId + "_" + this.javaClass.simpleName

    fun addTo(statusBar: StatusBar) {
        statusBar.addWidget(this, "before Position")
        statusBar.updateWidget(ID())
        floatingWidget.show()
    }

    fun updateOn(statusBar: StatusBar) {
        statusBar.updateWidget(ID())
        floatingWidget.show()
    }

    fun removeFrom(statusBar: StatusBar) {
        statusBar.removeWidget(ID())
        floatingWidget.hide()
    }

    interface Listener {
        fun onClick()
    }
}
